pkg/volume/iscsi/attacher.go

```diff
@@ -159,29 +160,67 @@ func (detacher *iscsiDetacher) UnmountDevice(deviceMountPath string) error {
 func (attacher *iscsiAttacher) volumeSpecToMounter(spec *volume.Spec, host volume.VolumeHost, pod *v1.Pod) (*iscsiDiskMounter, error) {
 	var secret map[string]string
 	var bkportal []string
-	iscsi, readOnly, err := getVolumeSource(spec)
+	readOnly, fsType, err := getISCSIVolumeInfo(spec)
 	if err != nil {
 		return nil, err
 	}
-	// Obtain secret for AttachDisk
-	if iscsi.SecretRef != nil && pod != nil {
-		if secret, err = volumeutil.GetSecretForPod(pod, iscsi.SecretRef.Name, host.GetKubeClient()); err != nil {
-			glog.Errorf("Couldn't get secret from %v/%v", pod.Namespace, iscsi.SecretRef)
+	if pod != nil {
```

Comment:
let's change this logic as little as possible, and do any unrelated optimizations as a followup

```
	// Obtain secret for AttachDisk
	if iscsi.SecretRef != nil && pod != nil {
		secretNamespace, secretName, err := getISCSISecretNamespaceAndName(spec, pod.Namespace)
		if err != nil {
			return nil, err
		}
		kubeClient := host.GetKubeClient()
		if kubeClient == nil {
			return nil, fmt.Errorf("cannot get kube client")
		}
		secretObj, err := kubeClient.Core().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
		if err != nil {
			return nil, fmt.Errorf("couldn't get secret %v/%v: %v", secretNamespace, secretName, err)
		}
		secret = map[string]string{}
		for name, data := range secretObj.Data {
			secret[name] = string(data)
		}
	}
```

Reply:
still outstanding... we weren't failing if chapDiscover or chapSession were set and there was no secretref before

Reply:
the secret is only meaningful if CHAP is enabled either for discovery or session or both. If none of these has enabled CHAP, we shouldn't fail if there is an invalid secret. 

Admittedly, in validation, the check only tests the case where CHAP is enabled but secret is missing.